File: leetcode-submissions/213_House_Robber_II.ts
function rob(nums: number[]): number {
    const dp1:Array<number> = new Array(nums.length+1).fill(-1);
     const dp2:Array<number> = new Array(nums.length+1).fill(-1);
    const n :number = nums.length;
    return Math.max(solve(nums , 0 , n-1,dp1), solve(nums , 1, n, dp2));
};

function solve(nums :number[], i :number , n :number , dp:Array<number>): number {
    if(i>=n){
        return 0 ;


    }
    if(dp[i]!=-1){
        return dp[i];
    }
    const take:number = nums[i]+solve(nums , i+2 , n, dp);
    const not_take:number = solve(nums , i+1, n, dp);
    return dp[i] =  Math.max(take , not_take);

};

File: leetcode-submissions/213_House_Robber_II.cpp
class Solution {
public:
    int t[101];
    int solve(vector<int>&nums, int index, int n ){
        if(index>n){
            return 0;

        }
        if(t[index]!=-1) return t[index];
        int left = solve(nums, index+2,n)+nums[index];
        int right = solve(nums,index+1,n);
        return t[index]=max(left, right);
    }
    int rob(vector<int>& nums) {
        memset(t,-1,sizeof(t));
        int from_0th_index = solve(nums, 0 ,nums.size()-2);
         memset(t,-1,sizeof(t));
        int from_1th_index= solve(nums, 1, nums.size()-1);
        return max(from_0th_index,from_1th_index);
    }
};